src/modules/python/pbs/v1/_pmi_cray.py

```diff
@@ -0,0 +1,424 @@
+            nids.append(first)
+        else:
+            nids.append(first + "-" + last)
+    return ",".join(nids), len(nidset)
+
+
+def spool_file(name):
+    # Form a path to the PBS spool directory with name as the last element.
+    home = _pbs_conf("PBS_HOME")
+    if home is None:
+        raise BackendError("PBS_HOME not found")
+    return os.path.join(home, "spool", name)
+
+
+def energy_file(job):
+    return spool_file("%s.energy" % job.id)
+
+
+def rur_file(job):
+    return spool_file("%s.rur" % job.id)
+
+
+def node_energy(id, nids, cnt):
+    # Return the result of running capmc get_node_energy_counter.
+    # The magic number of 15 seconds in the past is used because that
+    # is the most current value that can be expected from capmc.
+    if cnt == 0:
+        return None
+    cmd = "get_node_energy_counter --nids %s" % nids
+    ret = launch(id, cmd)
+    cntkey = "nid_count"
+    gotcnt = "<notset>"
+    if (ret is not None) and (cntkey in ret):
+        gotcnt = ret[cntkey]
+        if gotcnt == cnt:
+            return ret
+
+    pbs.logjobmsg(id, "node count %s, should be %d" % (str(gotcnt), cnt))
+    ret = launch(id, cmd)
+    gotcnt = "<notset>"
+    if (ret is not None) and (cntkey in ret):
+        gotcnt = ret[cntkey]
+        if gotcnt == cnt:
+            return ret
+
+    pbs.logjobmsg(id, "second query failed, node count %s, should be %d" %
+                      (str(gotcnt), cnt))
+    return None
+
+
+def job_energy(job, nids, cnt):
+    # Return energy counter from capmc.  Return None if no energy
+    # value is available.
+    energy = None
+    ret = node_energy(job.id, nids, cnt)
+    if ret is not None and "nodes" in ret:
+        energy = 0
+        for node in ret["nodes"]:
+            energy += node["energy_ctr"]
+        pbs.logjobmsg(job.id, "energy usage %dJ" % energy)
+    return energy
+
+
+class Pmi:
+
+    def __init__(self, pyhome=None):
+        pbs.logmsg(pbs.LOG_DEBUG, "Cray: init")
+
+    def _connect(self, endpoint, port):
+        pbs.logmsg(pbs.LOG_DEBUG, "Cray: connect")
+        return
+
+    def _disconnect(self):
+        pbs.logmsg(pbs.LOG_DEBUG, "Cray: disconnect")
+        return
+
+    def _get_usage(self, job):
+        pbs.logjobmsg(job.id, "Cray: get_usage")
+        try:
+            f = open(energy_file(job), "r")
+            start = int(f.read())
+            f.close()
+        except:
+            return None
+
+        e = pbs.event()
+        if e.type == pbs.EXECHOST_PERIODIC:
+            # This function will be called for each job in turn when
+            # running from a periodic hook.  Here we fill in some
+            # global variables just once and use the information
+            # for each job in turn.  Save the result of calling capmc
+            # for all running jobs in the variable ninfo.  Keep a
+            # dictionary with the job id's as keys holding a set
+            # of nid numbers.
+            global ninfo, nidarray
```

Comment:
Make these variables ninfo and nidarray as class level variables which can be access using classname. <variable_name> and acts as global to the class itself.

Reply:
Done.